packages/repository/src/repositories/relation.factory.ts

```diff
@@ -0,0 +1,164 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/example-todo
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {EntityCrudRepository} from '.';
+import {
+  Model,
+  RelationType,
+  AnyObject,
+  Options,
+  Entity,
+  Class,
+  FilterBuilder,
+  Filter,
+  WhereBuilder,
+  Where,
+  DataObject,
+} from '..';
+import {cloneDeep, isArray} from 'lodash';
+import {
+  HasManyEntityCrudRepository,
+  DefaultHasManyEntityCrudRepository,
+} from './relation.repository';
+
+export type RelationDefinitionBase = {
+  type: RelationType;
+  modelFrom: Class<Entity> | string;
+  keyTo: string;
+  keyFrom: string;
+};
+
+export interface HasManyDefinition extends RelationDefinitionBase {
+  type: RelationType.hasMany;
+}
+/**
+ * Enforces a constraint on a repository based on a relationship contract
+ * between models. Returns a relational repository that exposes applicable CRUD
+ * method APIs for the related target repository. For example, if a Customer model is
+ * related to an Order model via a HasMany relation, then, the relational
+ * repository returned by this method would be constrained by a Customer model
+ * instance's id(s).
+ *
+ * @param constraint The constraint to apply to the target repository. For
+ * example, {id: '5'}.
+ * @param relationMetadata The relation metadata used to used to describe the
+ * relationship and determine how to apply the constraint.
+ * @param targetRepository The repository which represents the target model of a
+ * relation attached to a datasource.
+ *
+ */
+export function constrainedRepositoryFactory<T extends Entity, ID>(
+  constraint: AnyObject,
+  relationMetadata: HasManyDefinition,
+  targetRepository: EntityCrudRepository<T, ID>,
+): HasManyEntityCrudRepository<T, ID> {
+  switch (relationMetadata.type) {
+    case RelationType.hasMany:
+      const fkConstraint: AnyObject = {};
+      fkConstraint[relationMetadata.keyTo] =
+        constraint[relationMetadata.keyFrom];
```

Comment:
This will not support composite keys.

I am proposing to simplify the initial implementation: Let `contstrainedRepositoryFactory` accept a single value of the source id (e.g. customerId). It's up to the consumer of this function to figure out that id. In a typical controller method (as we have in LB 3.x and envisioned for LB 4), this single id value is already provided by the caller.

```ts
export function relatedRepositoryFactory<SourceID, Target extends Entity, TargetId>(
 sourceModelId: SourceID,
 relationMetadata: HasManyDefinition,
 targetRepository: EntityCrudRepository<T, ID>,
) {
 // ...
 fkConstraint = {[relationMetadata.keyTo]: sourceModelId};
}
```

I understand this is not exactly the flexible "constrainedRepositoryFactory" design you we discussing with Raymond, but then your proposed "constrainedRepositoryFactory" implementation is not that much flexible either, since it requires a relation definition.

Reply:
Okay, although I was going for https://github.com/strongloop/loopback-next/pull/1342#discussion_r189272357, this would also be a good place to start :-).